base_layer/core/src/validation/block_validators/async_validator.rs

```diff
@@ -0,0 +1,432 @@
+            }
+
+            Ok(InputValidationData {
+                inputs,
+                aggregate_input_key,
+                commitment_sum,
+            })
+        })
+        .into()
+    }
+
+    fn start_output_validation(
+        &self,
+        header: &BlockHeader,
+        outputs: Vec<TransactionOutput>,
+    ) -> AbortOnDropJoinHandle<Result<OutputValidationData, ValidationError>> {
+        let height = header.height;
+        let num_outputs = outputs.len();
+        let concurrency = cmp::min(self.concurrency, num_outputs);
+        let queue = Arc::new(Mutex::new(outputs.into_iter().enumerate().collect::<Vec<_>>()));
+        let bypass_range_proof_verification = self.bypass_range_proof_verification;
+        if bypass_range_proof_verification {
+            info!(target: LOG_TARGET, "Range proof verification will be bypassed!")
+        }
+
+        debug!(
+            target: LOG_TARGET,
+            "Using {} worker(s) to validate #{} ({} output(s))", concurrency, height, num_outputs
+        );
+        let mut output_tasks = (0..concurrency)
+            .map(|_| {
+                let queue = queue.clone();
+                let range_proof_prover = self.factories.range_proof.clone();
+                let db = self.db.inner().clone();
+                task::spawn_blocking(move || {
+                    let db = db.db_read_access()?;
+                    let mut aggregate_sender_offset = PublicKey::default();
+                    let mut commitment_sum = Commitment::default();
+                    let mut outputs = Vec::new();
+                    let mut coinbase_index = None;
+                    while let Some((orig_idx, output)) = queue.lock().expect("lock poisoned").pop() {
+                        if output.is_coinbase() {
+                            if coinbase_index.is_some() {
+                                warn!(
+                                    target: LOG_TARGET,
+                                    "Block #{} failed to validate: more than one coinbase output", height
+                                );
+                                return Err(ValidationError::TransactionError(TransactionError::MoreThanOneCoinbase));
+                            }
+                            coinbase_index = Some(orig_idx);
+                        } else {
+                            // Lets gather the output public keys and hashes.
+                            // We should not count the coinbase tx here
+                            aggregate_sender_offset = aggregate_sender_offset + &output.sender_offset_public_key;
+                        }
+
+                        output.verify_metadata_signature()?;
+                        if bypass_range_proof_verification {
+                            output.verify_range_proof(&range_proof_prover)?;
+                        }
+
+                        helpers::check_not_duplicate_txo(&*db, &output)?;
```

Comment:
Also might be worth checking that the chain has not changed before we start this check

Reply:
Read lock prevents changes while the block is being validated

Reply:
(if you didn't know there are lines before the read lock where another thread can change the db) ;)